sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryQuerySource.java

```diff
@@ -76,61 +64,19 @@ private BigQueryQuerySource(
       Coder<T> coder,
       SerializableFunction<SchemaAndRecord, T> parseFn) {
     super(stepUuid, bqServices, coder, parseFn);
-    this.query = checkNotNull(query, "query");
-    this.flattenResults = checkNotNull(flattenResults, "flattenResults");
-    this.useLegacySql = checkNotNull(useLegacySql, "useLegacySql");
-    this.dryRunJobStats = new AtomicReference<>();
+    this.queryHelper = new BigQueryQueryHelper(query, flattenResults, useLegacySql);
   }
 
   @Override
   public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
     BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
-    return dryRunQueryIfNeeded(bqOptions).getTotalBytesProcessed();
+    return queryHelper.dryRunQueryIfNeeded(bqServices, bqOptions).getTotalBytesProcessed();
   }
 
   @Override
   protected TableReference getTableToExtract(BigQueryOptions bqOptions)
       throws IOException, InterruptedException {
-    // 1. Find the location of the query.
-    String location = null;
-    List<TableReference> referencedTables =
-        dryRunQueryIfNeeded(bqOptions).getQuery().getReferencedTables();
-    DatasetService tableService = bqServices.getDatasetService(bqOptions);
-    if (referencedTables != null && !referencedTables.isEmpty()) {
-      TableReference queryTable = referencedTables.get(0);
-      location = tableService.getTable(queryTable).getLocation();
-    }
-
-    // 2. Create the temporary dataset in the query location.
-    TableReference tableToExtract = createTempTableReference(
-        bqOptions.getProject(), createJobIdToken(bqOptions.getJobName(), stepUuid));
-
-    LOG.info("Creating temporary dataset {} for query results", tableToExtract.getDatasetId());
```

Comment:
Just to make sure, existing source will continue to be the default till the new version is stable, right ? (not sure why there's a large diff here).

Reply:
Yes, this is only for distribution to EAP customers. The reason there's a diff here is that this code is being refactored for consumption in the new read-from-query path, which doesn't provide a Source object.